### PR TITLE
Add environment_variables.txt file documentation

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/13669?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13669/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13669
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Follow-up for https://github.com/mlflow/mlflow/pull/13584

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: docs/source/models.rst [hunk 1]
```diff
@@ -103,6 +109,43 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
 
     model/dependencies
 
+Environment variables file
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+To track environment variables that are used during model inference, we generate ``environment_variables.txt`` file when logging a model.
+It **only contains names** of the environment variables that are used during model inference, **values are not stored**.
+Currently we only log the name if it includes any keywords in the following set:
```
Comment (harupy) [inline docs/source/models.rst:116]: can we replace we with mlflow?

File: docs/source/models.rst [hunk 2]
```diff
@@ -103,6 +109,43 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
...
+        "DATABRICKS_INSECURE",
+        "DATABRICKS_CLIENT_ID",
+        "DATABRICKS_CLIENT_SECRET",
+        "_DATABRICKS_WORKSPACE_HOST",
+        "_DATABRICKS_WORKSPACE_ID",
+        # dbconnect related
+        "SPARK_REMOTE",
+        "SPARK_LOCAL_REMOTE",
+    }
+
```
Comment (harupy) [inline docs/source/models.rst:138]: Can add an example model that uses environment variables?

File: docs/source/models.rst [hunk 3]
```diff
@@ -103,6 +109,43 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+To track environment variables that are used during model inference, MLflow generates ``environment_variables.txt`` file when logging a model.
+It **only contains names** of the environment variables that are used during model inference, **values are not stored**.
+Currently MLflow only logs the name if it includes any keywords in the following set:
+
+.. code-block:: python
+
+    RECORD_ENV_VAR_ALLOWLIST = {
+        # api key related
+        "API_KEY",
```
Comment (harupy) [inline docs/source/models.rst:122]:
```suggestion
        "API_KEY",  # e.g. OPENAI_API_KEY
```

File: docs/source/models.rst [hunk 4]
```diff
@@ -103,6 +109,43 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
 
     model/dependencies
 
+Environment variables file
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+To track environment variables that are used during model inference, MLflow generates ``environment_variables.txt`` file when logging a model.
```
Comment (harupy) [inline docs/source/models.rst:114]:
```suggestion
MLflow generates logs the environment variables used during model inference in ``environment_variables.txt`` file when logging a model.
```

File: docs/source/models.rst [hunk 5]
```diff
@@ -45,13 +45,16 @@ format. For example, :py:mod:`mlflow.sklearn` outputs models as follows:
 
 ::
 
-    # Directory written by mlflow.sklearn.save_model(model, "my_model")
-    my_model/
+    # Run `python examples/sklearn_logistic_regression/train.py` from mlflow directory
+    model/
```
Comment (harupy) [inline docs/source/models.rst:49]: I'd prefer the original description. If I was a user, I would wonder "where does examples/sklearn_logistic_regression/train.py come from?".

File: docs/source/models.rst [hunk 6]
```diff
@@ -45,13 +45,16 @@ format. For example, :py:mod:`mlflow.sklearn` outputs models as follows:
...
     ├── MLmodel
     ├── model.pkl
     ├── conda.yaml
     ├── python_env.yaml
-    └── requirements.txt
+    ├── requirements.txt
+    ├── input_example.json
+    ├── serving_input_example.json
```
Comment (harupy) [inline docs/source/models.rst:56]: `input_example.json` and `serving_input_example.json` are also optional, right?

File: docs/source/models.rst [hunk 7]
```diff
@@ -45,13 +45,16 @@ format. For example, :py:mod:`mlflow.sklearn` outputs models as follows:
...
+    └── environment_variables.txt (optional, only existing when environment variables are used)
```
Comment (harupy) [inline docs/source/models.rst:57]:
```suggestion
    └── environment_variables.txt (optional, only logged when environment variables are used during model inference)
```

We should be more precise.

File: docs/source/models.rst [hunk 8]
```diff
@@ -45,13 +45,16 @@ format. For example, :py:mod:`mlflow.sklearn` outputs models as follows:
 
 ::
 
-    # Directory written by mlflow.sklearn.save_model(model, "my_model")
-    my_model/
+    # Run `python examples/sklearn_logistic_regression/train.py` from mlflow directory
```
Comment (B-Step62) [inline docs/source/models.rst:48]:
```suggestion
    # Running `python examples/sklearn_logistic_regression/train.py` will create the following files under the `model` directory.
```

File: docs/source/models.rst (hunk 6 above)
Comment (B-Step62) [inline docs/source/models.rst:56]: Let's comment these are only created when an input example is passed.

File: docs/source/models.rst [hunk 9]
```diff
@@ -83,6 +86,9 @@ For environment recreation, we automatically log ``conda.yaml``, ``python_env.ya
 These files can then be used to reinstall dependencies using ``conda`` or ``virtualenv`` with ``pip``. Please see 
 :ref:`How MLflow Model Records Dependencies <how-mlflow-records-dependencies>` for more details about these files.
 
+If the model input example is provided when logging the model, additional files including ``input_example.json`` and ``serving_input_example.json`` are logged.
```
Comment (B-Step62) [inline docs/source/models.rst:89]:
```suggestion
If a model input example is provided when logging the model, two additional files ``input_example.json`` and ``serving_input_example.json`` are logged.
```

File: docs/source/models.rst [hunk 10]
```diff
@@ -103,6 +109,43 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
 
     model/dependencies
 
+Environment variables file
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+To track environment variables that are used during model inference, MLflow generates ``environment_variables.txt`` file when logging a model.
+It **only contains names** of the environment variables that are used during model inference, **values are not stored**.
```
Comment (B-Step62) [inline docs/source/models.rst:115]: nit: Can we make this `.. attention:`?

File: docs/source/models.rst [hunk 11]
```diff
@@ -103,6 +109,43 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
...
+Currently MLflow only logs the name if it includes any keywords in the following set:
```
Comment (B-Step62) [inline docs/source/models.rst:116]:
```suggestion
Currently MLflow only logs the environment variables whose name contains any of the following keywords:
```

File: docs/source/models.rst [hunk 12]
```diff
@@ -103,6 +109,43 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
...
+        # dbconnect related
+        "SPARK_REMOTE",
+        "SPARK_LOCAL_REMOTE",
+    }
+
+.. attention::
+    **Check the environment_variables.txt file before deploying the model to a serving endpoint**, as they might need 
+    to be set in the serving environment. Follow `this guidance <https://docs.databricks.com/en/machine-learning/model-serving/store-env-variable-model-serving.html#add-plain-text-environment-variables>`_
+    to set environment variables on a databricks serving endpoint. **Note that it is not guaranteed that all environment
+    variables tracked in the file are required for model inference.**
```
Comment (B-Step62) [inline docs/source/models.rst:143]:
```suggestion
    Before you deploy a model to a serving endpoint, **review the ``environment_variables.txt`` file** to ensure all necessary environment variables for model inference are set. Note that **not all environment 
    variables listed in the file are always required for model inference.** For detailed instructions on setting environment variables on a databricks serving endpoint, refer to `this guidance <https://docs.databricks.com/en/machine-learning/model-serving/store-env-variable-model-serving.html#add-plain-text-environment-variables>`_.
```

File: docs/source/models.rst [hunk 13]
```diff
@@ -103,6 +109,43 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
...
+
+.. note::
+    To disable logging environment variable names in model logging, set the environment variable
```
Comment (B-Step62) [inline docs/source/models.rst:146]:
```suggestion
    To disable this feature, set the environment variable 
```

**Review by B-Step62 — APPROVED** [review]
LGTM! All my comments are minor phrasing update.

File: docs/source/models.rst [hunk 14]
```diff
@@ -103,6 +110,75 @@ When logging a model, model metadata files (``MLmodel``, ``conda.yaml``, ``pytho
...
+        model_info = mlflow.pyfunc.log_model(
+            "model", python_model=MyModel(), input_example="data"
+        )
+
+Environment variable `TEST_API_KEY` is logged in the environment_variables.txt file
+
+.. figure:: ./_static/images/models/environment_variable_file_sample.png
+    :alt: The environment variables txt file example
+    :width: 80%
+    :align: center
```
Comment (harupy) [inline docs/source/models.rst:171]: Can we use a code block instead?

File: docs/source/models.rst [hunk 15]
```diff
@@ -41,17 +41,21 @@ MLmodel file
 ^^^^^^^^^^^^
 
 All of the flavors that a particular model supports are defined in its ``MLmodel`` file in YAML
-format. For example, :py:mod:`mlflow.sklearn` outputs models as follows:
+format. For example, running ``python examples/sklearn_logistic_regression/train.py`` from `MLflow repo <https://github.com/mlflow/mlflow>`_
```
Comment (harupy) [inline docs/source/models.rst:44]:
```suggestion
format. For example, running ``python examples/sklearn_logistic_regression/train.py`` from `MLflow repo <https://github.com/mlflow/mlflow/blob/master/examples/sklearn_logistic_regression/train.py>`_
```

can we link to the example?

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!

**Pull request merged**